File: lab3/quickSort.cpp
#include <iostream>
#include <time.h>
#include <vector>
#include <chrono>
using namespace std;

void insertionSort(vector<int> &array)
{
    for (size_t i = 1; i < array.size(); i++)
    {
        int key = array[i];
        int j = i - 1;

        while (j >= 0 && key < array[j])
        {
            array[j + 1] = array[j];
            j -= 1;
        }
        array[j + 1] = key;
    }
}
void insertionSort_desc(vector<int> &array)
{
    for (size_t i = 1; i < array.size(); i++)
    {
        int key = array[i];
        int j = i - 1;

        while (j >= 0 && key > array[j])
        {
            array[j + 1] = array[j];
            j -= 1;
        }
        array[j + 1] = key;
    }
}

int partition(vector<int> &array, int p, int r)
{
    int partitonElement = array[r];
    int i = p - 1;

    for (size_t j = p; j < r + 1; j++)
    {
        if (array[j] <= partitonElement)
        {
            i += 1;
            int temp = array[i];
            array[i] = array[j];
            array[j] = temp;
        }
    }
    if (i < r)
        return i;
    else
        return i - 1;
}

void quickSort(vector<int> &array, int p, int r)
{
    if (p < r)
    {
        int q = partition(array, p, r);
        quickSort(array, p, q);
        quickSort(array, q + 1, r);
    }
}

void quickSortModified(vector<int> &array, int p, int r, int c)
{
    if (p + c < r)
    {
        int q = partition(array, p, r);
        quickSortModified(array, p, q, c);
        quickSortModified(array, q + 1, r, c);
    }
}

void compareRandom()
{
    vector<int> array;

    for (size_t n = 1000; n < 100000; n += 2000)
    {
        for (size_t i = 0; i < n; i++)
        {
            array.push_back(rand() % 1001);
        }

        auto firstArray = array;
        auto secondArray = array;

        auto startFirst = chrono::high_resolution_clock::now();
        quickSort(firstArray, 0, firstArray.size() - 1);
        auto endFirst = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsedFirst = endFirst - startFirst;

        auto startSecond = chrono::high_resolution_clock::now();
        quickSortModified(secondArray, 0, secondArray.size() - 1, 100);
        insertionSort(array);
        auto endSecond = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsedSecond = endSecond - startSecond;

        printf("n: %6d\t time normal: %2.10f\t time modified: %2.10f\n", n, elapsedFirst.count(), elapsedSecond.count());

        array.clear();
    }
}

void compareDesc()
{
    vector<int> array;

    for (size_t n = 1000; n < 100000; n += 2000)
    {
        for (size_t i = 0; i < n; i++)
        {
            array.push_back(rand() % 1001);
        }
        insertionSort_desc(array);

        auto firstArray = array;
        auto secondArray = array;

        auto startFirst = chrono::high_resolution_clock::now();
        quickSort(firstArray, 0, firstArray.size() - 1);
        auto endFirst = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsedFirst = endFirst - startFirst;

        auto startSecond = chrono::high_resolution_clock::now();
        quickSortModified(secondArray, 0, secondArray.size() - 1, 100);
        insertionSort(array);
        auto endSecond = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsedSecond = endSecond - startSecond;

        printf("n: %6d\t time normal: %2.10f\t time modified: %2.10f\n", n, elapsedFirst.count(), elapsedSecond.count());

        array.clear();
    }
}

int main()
{

    srand(time(NULL));
    // compareRandom();
    // compareDesc();

    return 0;
}

// WARTOŚCI LOSOWE

// n:   1000        time normal: 0.0009998000       time modified: 0.0029986000
// n:   3000        time normal: 0.0009991000       time modified: 0.0240040000
// n:   5000        time normal: 0.0029996000       time modified: 0.0620065000
// n:   7000        time normal: 0.0030002000       time modified: 0.1180461000
// n:   9000        time normal: 0.0039990000       time modified: 0.1960466000
// n:  11000        time normal: 0.0050467000       time modified: 0.2939509000
// n:  13000        time normal: 0.0060000000       time modified: 0.4010003000
// n:  15000        time normal: 0.0080000000       time modified: 0.5459990000
// n:  17000        time normal: 0.0109996000       time modified: 0.7070007000
// n:  19000        time normal: 0.0089991000       time modified: 0.9090484000
// n:  21000        time normal: 0.0150461000       time modified: 1.1120004000
// n:  23000        time normal: 0.0159996000       time modified: 1.3500468000
// n:  25000        time normal: 0.0160471000       time modified: 1.5980005000
// n:  27000        time normal: 0.0169999000       time modified: 1.9920044000
// n:  29000        time normal: 0.0180467000       time modified: 2.4489524000
// n:  31000        time normal: 0.0200469000       time modified: 2.7500924000
// n:  33000        time normal: 0.0229980000       time modified: 2.9600118000
// n:  35000        time normal: 0.0230132000       time modified: 3.1859858000
// n:  37000        time normal: 0.0279996000       time modified: 3.8519993000
// n:  39000        time normal: 0.0260476000       time modified: 3.9040008000
// n:  41000        time normal: 0.0250476000       time modified: 4.3609999000
// n:  43000        time normal: 0.0349976000       time modified: 4.7830978000
// n:  45000        time normal: 0.0339983000       time modified: 5.5860008000
// n:  47000        time normal: 0.0410048000       time modified: 5.7599942000
// n:  49000        time normal: 0.0380006000       time modified: 6.2870103000
// n:  51000        time normal: 0.0410003000       time modified: 6.4460459000
// n:  53000        time normal: 0.0409873000       time modified: 7.3789997000
// n:  55000        time normal: 0.0460002000       time modified: 7.6191082000

// WARTOSCI MALEJĄCE

// n:   1000        time normal: 0.0040002000       time modified: 0.0100001000
// n:   3000        time normal: 0.0359508000       time modified: 0.0790460000
// n:   5000        time normal: 0.0749982000       time modified: 0.2090470000
// n:   7000        time normal: 0.1419962000       time modified: 0.3920484000
// n:   9000        time normal: 0.2099959000       time modified: 0.6449977000
// n:  11000        time normal: 0.3190523000       time modified: 1.0149526000
// n:  13000        time normal: 0.4029519000       time modified: 1.3230005000
// n:  15000        time normal: 0.5489999000       time modified: 1.7269988000
// n:  17000        time normal: 0.6849526000       time modified: 2.0620017000
// n:  19000        time normal: 0.7940050000       time modified: 2.5289947000
// n:  21000        time normal: 1.0129950000       time modified: 3.2069530000
// n:  23000        time normal: 1.2150026000       time modified: 3.9729976000
// n:  25000        time normal: 1.6321247000       time modified: 4.9900030000
// n:  27000        time normal: 1.6700419000       time modified: 5.3429530000
// n:  29000        time normal: 1.9959985000       time modified: 6.7400007000
// n:  31000        time normal: 2.1319994000       time modified: 6.9650070000
// n:  33000        time normal: 2.3669997000       time modified: 8.0279533000
// n:  35000        time normal: 2.5568513000       time modified: 8.9444538000
// n:  37000        time normal: 3.3709956000       time modified: 10.4440003000

File: lab3/quickSort.py
import random
import time
import sys
sys.setrecursionlimit(50000)

def insertionSort(array):
    for i in range(1, len(array)):
        key = array[i]
        j = i - 1

        while j >= 0 and key < array[j]:
            array[j+1] = array[j]
            j -= 1
        array[j + 1] = key

def insertionSort__desc(array):
    for i in range(1, len(array)):
        key = array[i]
        j = i - 1

        while j >= 0 and key > array[j]:
            array[j+1] = array[j]
            j -= 1
        array[j + 1] = key




def partition(array, p, r):
    partition_element = array[r]
    i = p - 1

    for j in range(p,r+1):
        if array[j] <= partition_element:
            i = i + 1
            temp = array[i]
            array[i] = array[j]
            array[j] = temp
    if i<r:
        return i
    else:
        return i - 1

def quickSortModified(array, p, r, c):
    if p + c<r:
        q = partition(array, p , r)
        quickSortModified(array, p, q, c)
        quickSortModified(array, q+1, r, c)

def quickSort(array, p, r):
    if p<r:
        q = partition(array, p , r)
        quickSort(array, p, q)
        quickSort(array, q+1, r)



# for n in range(1000, 97000, 2000):
#     array = []
#     flag = True
    
#     for i in range(n):
#         array.append(random.randint(0,1001))

#     start = time.perf_counter()
#     quickSort(array, 0 , len(array) - 1)
#     end = time.perf_counter()
#     timeElapsed = end - start
#     print("n: %1d \ttime: %2.10f" % (n, timeElapsed))


for n in range(1000, 97000, 2000):
    array = []
    flag = True
    
    for i in range(n):
        array.append(random.randint(0,1001))
    insertionSort__desc(array)
    start = time.perf_counter()
    quickSort(array, 0 , len(array) - 1)
    end = time.perf_counter()
    timeElapsed = end - start
    print("n: %1d \ttime: %2.10f" % (n, timeElapsed))


    # for i in range(0, len(array)-1):
    #     if array[i] > array[i+1]:
    #         print("ZLE!!!")
    #         flag = False
    #         break
    # if flag:
    #     print("POSORTOWANE!!!")


# WARTOSCI LOSOWE --- QUICKSORT NORMALNY

# n: 1000         time: 0.0047372000      
# n: 3000         time: 0.0207210000
# n: 5000         time: 0.0348198000
# n: 7000         time: 0.0572750000
# n: 9000         time: 0.0712465000
# n: 11000        time: 0.0770084000
# n: 13000        time: 0.1184848000
# n: 15000        time: 0.1472085000
# n: 17000        time: 0.1828012000
# n: 19000        time: 0.1928578000
# n: 21000        time: 0.2118892000
# n: 23000        time: 0.2443748000
# n: 25000        time: 0.2178280000
# n: 27000        time: 0.2939816000
# n: 29000        time: 0.2963727000
# n: 31000        time: 0.3056223000
# n: 33000        time: 0.3117778000
# n: 35000        time: 0.3462323000
# n: 37000        time: 0.3600484000
# n: 39000        time: 0.4273720000
# n: 41000        time: 0.5082267000
# n: 43000        time: 0.4587217000
# n: 45000        time: 0.5392993000
# n: 47000        time: 0.5398878000
# n: 49000        time: 0.6084368000
# n: 51000        time: 0.7566847000
# n: 53000        time: 0.8090041000
# n: 55000        time: 0.8418515000
# n: 57000        time: 0.8421982000
# n: 59000        time: 0.7584930000
# n: 61000        time: 0.8355577000
# n: 63000        time: 0.8661151000
# n: 65000        time: 0.8550830000
# n: 67000        time: 0.9603031000
# n: 69000        time: 0.9710779000
# n: 71000        time: 1.1051902000
# n: 73000        time: 1.0617699000
# n: 75000        time: 1.1579514000
# n: 77000        time: 1.1596250000
# n: 79000        time: 1.1914103000
# n: 81000        time: 1.2137445000
# n: 83000        time: 1.2967023000
# n: 85000        time: 1.6305914000
# n: 87000        time: 1.4671541000
# n: 89000        time: 1.6355332000
# n: 91000        time: 1.8983558000
# n: 93000        time: 1.7135441000
# n: 95000        time: 1.8507459000

# WARTOŚCI LOSOWE --- QUICKSORT ZMODYFIKOWANY
# n: 1000         time: 0.0061274000
# n: 3000         time: 0.0169511000
# n: 5000         time: 0.0291312000
# n: 7000         time: 0.0410995000
# n: 9000         time: 0.0542136000
# n: 11000        time: 0.0661670000
# n: 13000        time: 0.0783200000
# n: 15000        time: 0.0884439000
# n: 17000        time: 0.1068893000
# n: 19000        time: 0.1237544000
# n: 21000        time: 0.1190097000
# n: 23000        time: 0.1320051000
# n: 25000        time: 0.1339326000
# n: 27000        time: 0.1571298000
# n: 29000        time: 0.1606354000
# n: 31000        time: 0.1884420000
# n: 33000        time: 0.1658900000
# n: 35000        time: 0.1859972000
# n: 37000        time: 0.2127205000
# n: 39000        time: 0.2123857000
# n: 41000        time: 0.2222726000
# n: 43000        time: 0.2493411000
# n: 45000        time: 0.2555687000
# n: 47000        time: 0.2338322000
# n: 49000        time: 0.2518408000
# n: 51000        time: 0.2439962000
# n: 53000        time: 0.2407833000
# n: 55000        time: 0.2404226000
# n: 57000        time: 0.2599131000
# n: 59000        time: 0.2515338000
# n: 61000        time: 0.3052480000
# n: 63000        time: 0.2793207000
# n: 65000        time: 0.2880571000
# n: 67000        time: 0.3081252000
# n: 69000        time: 0.3026281000
# n: 71000        time: 0.3221743000
# n: 73000        time: 0.3139365000
# n: 75000        time: 0.3457785000
# n: 77000        time: 0.3663756000
# n: 79000        time: 0.3438824000
# n: 81000        time: 0.3481167000
# n: 83000        time: 0.3510895000
# n: 85000        time: 0.3753278000
# n: 87000        time: 0.4296606000
# n: 89000        time: 0.4137193000
# n: 91000        time: 0.4225715000
# n: 93000        time: 0.4122823000
# n: 95000        time: 0.5149074000